src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+    ASSERT_EQ(installed_versions.size(), 1);
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    // why 0, this is secondary that was succesfully installed, so it should 1/true ???
+    ASSERT_EQ(current_version, 0);
+  }
+
+  {
+    // now try to finalize the previously installed updates with enabled failure injection
+    fiu_init(0);
+    fiu_enable("fake_install_finalization_failure", 1, nullptr, 0);
+    http_server_mock->clearReportEvents();
+
+    Aktualizr aktualizr(conf, storage, http_server_mock);
+    EventHandler event_hdlr(aktualizr);
+    aktualizr.Initialize();
+    aktualizr.UptaneCycle();
+
+    fiu_disable("fake_install_finalization_failure");
+
+    result::UpdateCheck update_res = aktualizr.CheckUpdates().get();
+    // Is it correct expectations ?????
+    // it fails because it sees that aktualizr got higher version (2) than the one that server returns 1
+    // moreover it takes version out of the stored metadata but not from successfully installed target
+    // see sotauptaneclient.cc::413
+    ASSERT_EQ(update_res.status, result::UpdateStatus::kError);
+    ASSERT_FALSE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents({"UpdateCheckComplete"}));
+    ASSERT_TRUE(http_server_mock->checkRecievedReports({"EcuInstallationCompleted"}));
+    ASSERT_FALSE(http_server_mock->wasInstallSuccessfull());
+
+    data::InstallationResult dev_installation_res;
+    std::string report;
+    std::string correlation_id;
+
+    // `device_installation_result` and `ecu_installation_results` are cleared
+    // at finalizeAfterReboot()->putManifestSimple() once a device manifest is successfully sent to a server
+    ASSERT_FALSE(storage->loadDeviceInstallationResult(&dev_installation_res, &report, &correlation_id));
+
+    // it's used to return `true` even if there is no any record in DB
+    // of the uptane cycle just after sending manifest
+    // made it consistent with loadDeviceInstallationResult
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_FALSE(storage->loadEcuInstallationResults(&ecu_installation_res));
+
+    // verify currently installed version
+    std::vector<Uptane::Target> installed_versions;
+    size_t current_version{0};  // why it's not boolean
+    size_t pending_version{0};  // why it's not boolean
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    // why valid , BTW this `valid` value is neither read from DB nor calculated and it's true by default ???
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    current_version = 0;  // why it's not boolean
+    pending_version = 0;  // why it's not boolean
```

Comment:
I thought this was an SQL limitation or something like that.

Reply:
solved & understood.